# Parte 2: Crear Dataset

### Importación de imagenes y rutas

In [1]:
import cv2
import numpy as np
import os


# Ruta base donde se encuentran las imágenes y donde se guardarán las subimágenes
basepath = r'C:\Users\Daniel Calderon\Desktop\2024-2 Poli\Vision Con IA\Practica3_(Parcial)'
path1 = '\Parte_1'
# Rutas de las imágenes filtradas
ruta1 = os.path.join(basepath + path1, 'imagen1_filtrada.png')
ruta2 = os.path.join(basepath + path1, 'imagen2_filtrada.png')
ruta3 = os.path.join(basepath + path1, 'imagen3_filtrada.png')


### Función para dividir y guardar las subimágenes de 512x512 en una sola carpeta

In [2]:
# Función para dividir y guardar las subimágenes de 512x512 en una sola carpeta
def dividir_y_guardar(imagen, nombre_base, directorio_base):
    alto, ancho = imagen.shape[:2]
    size = 512
    contador = 0

    # Crear directorio para las subimágenes si no existe
    if not os.path.exists(directorio_base):
        os.makedirs(directorio_base)

    for i in range(0, alto, size):
        for j in range(0, ancho, size):
            subimagen = imagen[i:i+size, j:j+size]
            # Solo guardar si la subimagen tiene el tamaño correcto
            if subimagen.shape[0] == size and subimagen.shape[1] == size:
                nombre_subimagen = f'{nombre_base}_{contador}.png'
                cv.imwrite(os.path.join(directorio_base, nombre_subimagen), subimagen)
                contador += 1

### Cargar imágenes filtradas

In [3]:
# Cargar imágenes filtradas
imagen1 = cv.imread(ruta1)
imagen2 = cv.imread(ruta2)
imagen3 = cv.imread(ruta3)

print('Tamaños originales:')
print('El tamaño de la imagen1 es', imagen1.shape, 'pixeles')
print('El tamaño de la imagen2 es', imagen2.shape, 'pixeles')
print('El tamaño de la imagen3 es', imagen3.shape, 'pixeles')

Tamaños originales:
El tamaño de la imagen1 es (11363, 25202, 3) pixeles
El tamaño de la imagen2 es (11364, 25204, 3) pixeles
El tamaño de la imagen3 es (11347, 25204, 3) pixeles


### Redimensionar imágenes a múltiplos de 512

In [4]:
# Función para redimensionar la imagen a múltiplos exactos de 512
def ajustar_a_multiplo_de_512(imagen):
    alto, ancho = imagen.shape[:2]
    nuevo_alto = (alto // 512) * 512
    nuevo_ancho = (ancho // 512) * 512
    return cv.resize(imagen, (nuevo_ancho, nuevo_alto))

### Imprimir tamaños redimensionados

In [5]:
# Redimensionar imágenes a múltiplos de 512
imagen1 = ajustar_a_multiplo_de_512(imagen1)
imagen2 = ajustar_a_multiplo_de_512(imagen2)
imagen3 = ajustar_a_multiplo_de_512(imagen3)

print('Tamaños ajustados:')
print('El tamaño de la imagen1 es', imagen1.shape, 'pixeles')
print('El tamaño de la imagen2 es', imagen2.shape, 'pixeles')
print('El tamaño de la imagen3 es', imagen3.shape, 'pixeles')

Tamaños ajustados:
El tamaño de la imagen1 es (11264, 25088, 3) pixeles
El tamaño de la imagen2 es (11264, 25088, 3) pixeles
El tamaño de la imagen3 es (11264, 25088, 3) pixeles


### Dividir y guardar las subimgenes a una sola carpeta en esta ruta

In [6]:
# Definir un único directorio donde se guardarán todas las subimágenes
path2 = '\Parte_2'
directorio_subimagenes = os.path.join(basepath + path2, 'Subimagenes')

# Dividir y guardar las subimágenes de las tres imágenes en la misma carpeta
dividir_y_guardar(imagen1, 'imagen1', directorio_subimagenes)
dividir_y_guardar(imagen2, 'imagen2', directorio_subimagenes)
dividir_y_guardar(imagen3, 'imagen3', directorio_subimagenes)

print("Subimágenes guardadas correctamente en una sola carpeta.")

Subimágenes guardadas correctamente en una sola carpeta.


# Explicación del Código: Dividir y Guardar Subimágenes

## Descripción General

Este programa carga tres imágenes filtradas, las redimensiona a múltiplos de 512x512 píxeles, y luego las divide en subimágenes de ese tamaño, guardando las subimágenes resultantes en una carpeta específica. El objetivo es procesar imágenes de gran tamaño y dividirlas en partes más manejables para su posterior análisis.

## Librerías Utilizadas

- **OpenCV (`cv2`)**: Para cargar, redimensionar, dividir y guardar las imágenes.
- **os**: Para gestionar rutas de archivos y directorios.
- **NumPy (`np`)**: Para operaciones de matrices de imágenes (utilizada implícitamente a través de OpenCV).

## Flujo del Programa

### 1. Definición de Rutas

El programa define las rutas para las imágenes filtradas y establece una carpeta base donde se guardarán las subimágenes.

- La variable `basepath` contiene la ruta base del proyecto.
- Las imágenes filtradas se encuentran en la carpeta `Parte_1` y las subimágenes se guardarán en `Parte_2`.

### 2. Función `dividir_y_guardar`

Esta función es la responsable de dividir las imágenes en subimágenes de 512x512 píxeles y guardarlas en una carpeta.

- **Parámetros**:
  - `imagen`: Imagen cargada en formato OpenCV.
  - `nombre_base`: Nombre base para los archivos de subimágenes.
  - `directorio_base`: Directorio donde se guardarán las subimágenes.

- **Proceso**:
  1. Verifica si el directorio de destino existe, y si no, lo crea.
  2. Recorre la imagen en bloques de 512x512 píxeles y guarda cada subimagen con un nombre incrementado.

### 3. Carga de Imágenes Filtradas

Se cargan tres imágenes filtradas usando `cv2.imread()`. Estas imágenes están en formato `.png`.

- La función `cv2.imread()` retorna la imagen en formato BGR (Blue-Green-Red).
- Se imprimen los tamaños originales de las imágenes usando `imagen.shape`.

### 4. Función `ajustar_a_multiplo_de_512`

Dado que las imágenes pueden no tener un tamaño exacto en múltiplos de 512x512, esta función ajusta el tamaño de las imágenes para que sus dimensiones sean múltiplos exactos de 512.

- **Proceso**:
  - Se calcula el tamaño más cercano en múltiplos de 512 tanto en altura como en ancho.
  - Se redimensiona la imagen usando `cv2.resize()` para que sus nuevas dimensiones sean divisibles por 512.

### 5. Redimensionar las Imágenes

Cada una de las tres imágenes cargadas se redimensiona utilizando la función `ajustar_a_multiplo_de_512()`.

### 6. Guardado de Subimágenes

Luego de ajustar el tamaño, cada imagen se divide en bloques de 512x512 píxeles y se guardan en una carpeta especificada. Las subimágenes se guardan en el directorio `Subimagenes`.

### 7. Resultados

Finalmente, el programa imprime un mensaje indicando que las subimágenes han sido guardadas correctamente.

## Mensajes de Consola

Durante la ejecución del programa, se imprimen los tamaños de las imágenes originales y ajustadas para verificar el proceso de redimensionamiento.

### Ejemplo de salida:
```bash
Tamaños originales:
El tamaño de la imagen1 es (11363, 25202, 3) pixeles
El tamaño de la imagen2 es (11364, 25204, 3) pixeles
El tamaño de la imagen3 es (11347, 25204, 3) pixeles

Tamaños ajustados:
El tamaño de la imagen1 es (11264, 25088, 3) pixeles
El tamaño de la imagen2 es (11264, 25088, 3) pixeles
El tamaño de la imagen3 es (11264, 25088, 3) pixeles

Subimágenes guardadas correctamente en una sola carpeta.
